rs/registry/admin/bin/main.rs

```diff
@@ -4018,33 +4020,69 @@ async fn main() {
             let contents = read_to_string(secret_key_path).expect("Could not read key file");
-            let sig_keys = SigKeys::from_pem(&contents).expect("Failed to parse pem file");
-            Sender::SigKeys(sig_keys)
+
+            // Try to parse as Ed25519 (BasicIdentity) first
+            if let Ok(identity) = BasicIdentity::from_pem(contents.as_bytes()) {
+                Arc::new(identity)
+            }
+            // Try to parse as Secp256k1 with RFC 5915 format (EC PRIVATE KEY)
+            else if let Ok(identity) = Secp256k1Identity::from_pem(contents.as_bytes()) {
+                Arc::new(identity)
+            } else {
+                panic!(
+                    "Failed to parse PEM file. Tried Ed25519 (BasicIdentity), Secp256k1 RFC5915, and Secp256k1 PKCS#8 formats."
+                );
+            }
         } else if opts.use_hsm {
-            make_hsm_sender(
-                &opts.hsm_slot.expect(
-                    "HSM slot must also be provided for --use-hsm; use --hsm-slot or see --help.",
-                ),
-                &opts.hsm_key_id.expect(
-                    "HSM key ID must also be provided for --use-hsm; use --key-id or see --help.",
-                ),
-                &opts.hsm_pin.expect(
-                    "HSM pin must also be provided for --use-hsm; use --pin or see --help.",
-                ),
+            // Parse slot as usize (ic-identity-hsm expects usize, not string)
+            let hsm_slot = opts
+                .hsm_slot
+                .as_ref()
+                .expect("HSM slot must be provided for --use-hsm")
+                .parse::<usize>()
+                .expect("HSM slot must be a valid number");
+
+            let hsm_key_id = opts
+                .hsm_key_id
+                .as_ref()
+                .expect("HSM key ID must be provided for --use-hsm");
+
+            let hsm_pin = opts
+                .hsm_pin
+                .clone()
+                .expect("HSM pin must be provided for --use-hsm");
+
+            // Use PKCS#11 library path from environment, or auto-detect from pkcs11-tool
+            let pkcs11_lib_path = std::env::var("PKCS11_LIB_PATH").unwrap_or_else(|_| {
+                find_opensc_pkcs11_lib()
+                    .map(|p| p.to_string_lossy().to_string())
+                    .expect(
+                        "Could not find pkcs11 library. Please set PKCS11_LIB_PATH, or \
+                            make sure pkcs11-tool is on your PATH.",
+                    )
+            });
+
+            Arc::new(
+                HardwareIdentity::new(pkcs11_lib_path, hsm_slot, hsm_key_id, move || {
+                    Ok(hsm_pin.clone())
+                })
+                .expect("Failed to create HardwareIdentity from HSM"),
             )
         } else {
-            Sender::Anonymous
+            Arc::new(AnonymousIdentity)
         }
     } else {
-        Sender::Anonymous
+        Arc::new(AnonymousIdentity)
     };
```

Comment:
Ofc, this issue already existed, so do not fix in this PR, but I just have to point out how this is totally backwards. Look at how far away these are from the conditions that trigger them. It's kind of insane. It should be more like this:

```
let identity = new_identity_from_opts_or_panic(&opts);

...

fn new_identity_from_opts_or_panic(&opts) -> impl Identity {
    if let Some(secret_key_pem) = opts.secret_key_pem.as_ref() {
        return new_pem_identity_or_panic(secret_key_pem);
    }

    if opts.use_hsm {
        return new_hsm_identity_or_panic(
            &opts.hsm_slot,
            ...
        );
    }

    AnonymousIdentity
}
```